cpp/include/cuml/common/logger.hpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stdarg.h>
+#include <memory>
+#include <string>
+
+namespace spdlog {
+class logger;
+};
+
+namespace ML {
+
+/**
+ * @defgroup CStringFormat Expand a C-style format string
+ *
+ * @brief Expands C-style formatted string into std::string
+ *
+ * @param[in] fmt format string
+ * @param[in] vl  respective values for each of format modifiers in the string
+ *
+ * @return the expanded `std::string`
+ *
+ * @{
+ */
+std::string format(const char* fmt, va_list& vl);
+std::string format(const char* fmt, ...);
+/** @} */
+
+/**
+ * @defgroup CumlLogLevels Logging levels used in cuML
+ *
+ * @note exactly match the corresponding ones in spdlog for wrapping purposes
+ *
+ * @{
+ */
+#define CUML_LEVEL_TRACE 0
+#define CUML_LEVEL_DEBUG 1
+#define CUML_LEVEL_INFO 2
+#define CUML_LEVEL_WARN 3
+#define CUML_LEVEL_ERROR 4
+#define CUML_LEVEL_CRITICAL 5
+#define CUML_LEVEL_OFF 6
+/** @} */
+
+#if !defined(CUML_ACTIVE_LEVEL)
+#define CUML_ACTIVE_LEVEL CUML_LEVEL_INFO
+#endif
+
+/**
+ * @brief The main Logging class for cuML library.
+ *
+ * This class acts as a thin wrapper over the underlying `spdlog` interface. The
+ * design is done in this way in order to avoid us having to also ship `spdlog`
+ * header files in our installation.
+ *
+ * @todo This currently only supports logging to stdout. Need to add support in
```

Comment:
Let's add a Github issue and reference it here so that this todo doesn't get lost in the codebase.

Reply:
Sure Corey. Filed issue #2046 for this one. Will also update this todo marker with the issue ID as well.